README.md
# aurelia-http-client

[![npm Version](https://img.shields.io/npm/v/aurelia-http-client.svg)](https://www.npmjs.com/package/aurelia-http-client)
[![ZenHub](https://raw.githubusercontent.com/ZenHubIO/support/master/zenhub-badge.png)](https://zenhub.io)
[![Join the chat at https://gitter.im/aurelia/discuss](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/aurelia/discuss?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![CircleCI](https://circleci.com/gh/aurelia/http-client.svg?style=shield)](https://circleci.com/gh/aurelia/http-client)

This library is part of the [Aurelia](http://www.aurelia.io/) platform and contains a simple, restful, message-based wrapper around XMLHttpRequest.

To keep up to date on [Aurelia](http://www.aurelia.io/), please visit and subscribe to [the official blog](http://blog.aurelia.io/) and [our email list](http://eepurl.com/ces50j). We also invite you to [follow us on twitter](https://twitter.com/aureliaeffect). If you have questions look around our [Discourse forums](https://discourse.aurelia.io/), chat in our [community on Gitter](https://gitter.im/aurelia/discuss) or use [stack overflow](http://stackoverflow.com/search?q=aurelia). Documentation can be found [in our developer hub](http://aurelia.io/docs). If you would like to have deeper insight into our development process, please install the [ZenHub](https://zenhub.io) Chrome or Firefox Extension and visit any of our repository's boards.

## Documentation

If possible, prefer to use the `aurelia-fetch-client` instead of this library. You can read documentation on the fetch client [here](http://aurelia.io/docs/plugins/http-services#options) along with documentation on this library. If you would like to help improve this documentation, the source for these docs can be found in the `aurelia-fetch-client` repo's doc folder.

## Platform Support

This library can be used in the **browser** only.

## Building The Code

To build the code, follow these steps.

1. Ensure that [NodeJS](http://nodejs.org/) is installed. This provides the platform on which the build tooling runs.
2. From the project folder, execute the following command:

  ```shell
  npm install
  ```
3. Ensure that [Gulp](http://gulpjs.com/) is installed. If you need to install it, use the following command:

  ```shell
  npm install -g gulp
  ```
4. To build the code, you can now run:

  ```shell
  gulp build
  ```
5. You will find the compiled code in the `dist` folder, available in three module formats: AMD, CommonJS and ES6.

6. See `gulpfile.js` for other tasks related to generating the docs and linting.

## Running The Tests

To run the unit tests, first ensure that you have followed the steps above in order to install all dependencies and successfully build the library. Once you have done that, proceed with these additional steps:

1. Ensure that the [Karma](http://karma-runner.github.io/) CLI is installed. If you need to install it, use the following command:

  ```shell
  npm install -g karma-cli
  ```
2. Ensure that [jspm](http://jspm.io/) is installed. If you need to install it, use the following commnand:

  ```shell
  npm install -g jspm
  ```
3. Install the client-side dependencies with jspm:

  ```shell
  jspm install
  ```

4. You can now run the tests with this command:

  ```shell
  karma start
  ```


doc/CHANGELOG.md
<a name="1.3.2"></a>
## [1.3.2](https://github.com/aurelia/http-client/compare/1.3.0...1.3.2) (2019-03-27)



<a name="1.3.1"></a>
## [1.3.1](https://github.com/aurelia/http-client/compare/1.3.0...1.3.1) (2019-02-04)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/aurelia/http-client/compare/1.2.1...1.3.0) (2018-06-14)

* Fix unit tests.
* Enable error responses to be shaped like Error objects, addressing some issues with BlueBird.

<a name="1.2.1"></a>
## [1.2.1](https://github.com/aurelia/http-client/compare/1.2.0...1.2.1) (2017-10-07)


### Bug Fixes

* **http-client:** don't call withParams if there are no params ([2cd076a](https://github.com/aurelia/http-client/commit/2cd076a))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/aurelia/http-client/compare/1.1.1...1.2.0) (2017-10-01)

### Features

* Added support for traditional query string parameter construction.

<a name="1.1.1"></a>
## [1.1.1](https://github.com/aurelia/http-client/compare/1.1.0...v1.1.1) (2017-04-05)


### Bug Fixes

* **xhr-transformers:** fix sending ArrayBuffer type ([4d957f8](https://github.com/aurelia/http-client/commit/4d957f8)), closes [#161](https://github.com/aurelia/http-client/issues/161)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/aurelia/http-client/compare/1.0.4...v1.1.0) (2017-03-23)


### Features

* **request-builder:** add download progress callback ([a53869b](https://github.com/aurelia/http-client/commit/a53869b)), closes [#163](https://github.com/aurelia/http-client/issues/163)
* **request-builder:** add download progress callback ([73165dd](https://github.com/aurelia/http-client/commit/73165dd))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/aurelia/http-client/compare/1.0.3...v1.0.4) (2017-01-06)

### Bug Fixes

* Improve dts files.

<a name="1.0.3"></a>
## [1.0.3](https://github.com/aurelia/http-client/compare/1.0.2...v1.0.3) (2016-11-08)


### Bug Fixes

* **http-response-message:** handle empty responses ([bae5188](https://github.com/aurelia/http-client/commit/bae5188))



## 1.0.2

### Bug Fixes

* **headers:** #152 - Made header treatment case insensitive

<a name="1.0.1"></a>
## [1.0.1](https://github.com/aurelia/http-client/compare/1.0.0...v1.0.1) (2016-10-06)


### Bug Fixes

* Add responseType field in HttpResponseMessage ([2c0f437](https://github.com/aurelia/http-client/commit/2c0f437)), closes [#149](https://github.com/aurelia/http-client/issues/149)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/aurelia/http-client/compare/1.0.0-rc.1.0.1...v1.0.0) (2016-07-27)


### Bug Fixes

* **request-message-processor:** do not send undefined message.content ([ede329f](https://github.com/aurelia/http-client/commit/ede329f))



<a name="1.0.0-rc.1.0.1"></a>
# [1.0.0-rc.1.0.1](https://github.com/aurelia/http-client/compare/1.0.0-rc.1.0.0...v1.0.0-rc.1.0.1) (2016-07-12)



<a name="1.0.0-rc.1.0.0"></a>
# [1.0.0-rc.1.0.0](https://github.com/aurelia/http-client/compare/1.0.0-beta.2.0.1...v1.0.0-rc.1.0.0) (2016-06-22)



### 1.0.0-beta.1.2.2 (2016-05-24)


### 1.0.0-beta.1.2.1 (2016-05-10)


### 1.0.0-beta.1.2.0 (2016-03-22)

* Update to Babel 6

### 1.0.0-beta.1.1.2 (2016-03-01)


#### Bug Fixes

* **all:** remove core-js dependency ([9fbd7413](http://github.com/aurelia/http-client/commit/9fbd7413058b53c2b98e51b952e869ca0b1c4b26))
* **http-client:** declare replacer property ([0080eaae](http://github.com/aurelia/http-client/commit/0080eaae426073a065bbf5f3d87b05f225c2a65c))


### 1.0.0-beta.1.1.1 (2016-02-08)


#### Bug Fixes

* **http-response:** status code type ([2afe081e](http://github.com/aurelia/http-client/commit/2afe081e36bfbaedab1541291cfd4ea492b69ca8), closes [#114](http://github.com/aurelia/http-client/issues/114))


## 1.0.0-beta.1.1.0 (2016-01-29)


#### Features

* **package:** update metadata for jspm; core-js; deps ([9b0e4272](http://github.com/aurelia/http-client/commit/9b0e4272ac4b784642d49aba87f206a949023868))


### 1.0.0-beta.1.0.1 (2016-01-08)

#### Bug Fixes

* **http-client:** declare headers property ([e64dba37](http://github.com/aurelia/http-client/commit/e64dba37c9539091fcc0e71a07e49aec60dda53f))


### 1.0.0-beta.1 (2015-11-16)


## 0.13.0 (2015-11-09)


#### Features

* **headers:** case-insensitve has ([494918d5](http://github.com/aurelia/http-client/commit/494918d54784ee0c214e21f60546c17a533f9bcf), closes [#92](http://github.com/aurelia/http-client/issues/92))
* **request-builder:** enable opening xhr with login info ([a9adff68](http://github.com/aurelia/http-client/commit/a9adff68a58813bd24c2f35b465929a2bc469ebc), closes [#52](http://github.com/aurelia/http-client/issues/52))


## 0.12.0 (2015-10-13)


#### Bug Fixes

* ***:** switch http client to use path.join ([6954c4fe](http://github.com/aurelia/http-client/commit/6954c4fe74de49739cdab92e3753f418136d3b7f))
* **all:**
  * add types for promises of send methods ([1911331d](http://github.com/aurelia/http-client/commit/1911331de6ca0325ceedcbff2db7a160110d0595))
  * update compiler and fix core-js references ([c0334447](http://github.com/aurelia/http-client/commit/c0334447b4a2f794b0428a54f5e43eebd663780c))
  * errors should still return HttpResponseMessage ([43f3cf4f](http://github.com/aurelia/http-client/commit/43f3cf4ffdd9690bebb99cffe699eacf3477de94))
* **bower:** correct semver ranges ([892535e1](http://github.com/aurelia/http-client/commit/892535e1ecea50c0cabd6fa9be25ebb1249632fd))
* **build:**
  * update linting, testing and tools ([f808d92c](http://github.com/aurelia/http-client/commit/f808d92c5140e60783f298dac1a8552048bb47dc))
  * correct order of files for concat ([5a6418de](http://github.com/aurelia/http-client/commit/5a6418de752733167dca716bb5aefac682286de4))
  * add missing bower bump ([59e8a734](http://github.com/aurelia/http-client/commit/59e8a734b555440ee26737040778dd18b2189d86))
* **event:** switch from hook to event ([01628f42](http://github.com/aurelia/http-client/commit/01628f42884c899ea5b6cc64ac118e5c8f9413c1))
* **http-client:**

karma.conf.js
// Karma configuration
// Generated on Fri Dec 05 2014 16:49:29 GMT-0500 (EST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jspm', 'source-map-support', 'jasmine-ajax', 'jasmine'],

    jspm: {
      // Edit this to your needs
      loadFiles: ['test/**/*.js'],
      serveFiles : ['src/**/*.js']
    },


    // list of files / patterns to load in the browser
    files: [],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.js': ['babel'],
      'src/**/*.js': ['babel']
    },
    'babelPreprocessor': {
      options: {
        sourceMap: 'inline',
        presets: [ 'es2015-loose', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      }
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
    
  });
};
